### PR TITLE
Generate a copy of ceph.conf locally

### PR DESCRIPTION
**Pull request by haphan**
Refers to #2697 

This change creates a copy of `ceph.conf` in ansible server.

Room for improvement:  a toggle flag to control this behavior.

@leseb What do you think?

Signed-off-by: Ha Phan <thanhha.work@gmail.com>

### CODE REVIEW

**Comment by haphan** [comment]
done :)

File: roles/ceph-config/tasks/main.yml [hunk 1]
```diff
@@ -34,6 +34,9 @@
       dest: "{{ fetch_directory }}/{{ fsid }}/etc/ceph/{{ cluster }}.conf"
       config_overrides: "{{ ceph_conf_overrides }}"
       config_type: ini
+    when:
+      - inventory_hostname in groups.get(mon_group_name, [])
```
Comment (guits) [inline roles/ceph-config/tasks/main.yml]: `inventory_hostname == groups[mon_group_name][0]`

**Comment by guits** [comment]
if the `when` is limiting to the last mon only, the `run_once: true` is not needed then

@leseb @haphan

**Comment by haphan** [comment]
done

**Comment by guits** [comment]
just to clarify:

when limiting to a particular mon, you shouldn't use `run_once: true`

eg:

```
when:
  - inventory_hostname == groups[mon_group_name][-1:]
run_once: true
```

if the first node played is the first mon in `mon_group_name` this will make the task skipped and only run once, it means the task won't be ever played

the better approach if you want to use `run_once: true` and you don't need to care on which mon the task will be executed is this approach:

```
when:
  - inventory_hostname in groups[mon_group_name]
run_once: true
```

**Comment by guits** [comment]
@haphan sorry for these multiple request changes, but would you please update your patch to fit the 2nd approach I've mentioned in my previous comment?
Thanks for your patience!

**Comment by haphan** [comment]
@guits I've changed back to 2nd approach. 

@andymcc In my case, Ceph cluster is deployed from a container e.g. `haphan/ceph-installer`. This container role is similar to ansible server; containing this `ceph-ansible` playbook along with python dependencies, etc..

Now I want to extend this container so that it can manage the Ceph cluster like create pools, managing osds, etc.. And for that, besides keyrings which we already have locally, the container also needs access to `ceph.conf`

I am still experimenting this approach and until now it works for me. Do share your thoughts.

**Comment by haphan** [comment]
Any chance to backport this feature to 3.0 branch?

**Comment by guits** [comment]
@haphan this looks more like a feature/enhancement than a critical bug fixing, therefore, I tend to say we can't backport this to stable.

**Comment by guits** [comment]
jenkins test luminous-ansible2.4-centos7_cluster

**Comment by andymcc** [comment]
Does setting your container up as a client not work? Or you would rather not use the client role for some reason?
I'm guessing once the ceph.conf is in your fetch directory you will either have to move it or do some other action in order for it to be useful?

**Comment by haphan** [comment]
@andymcc client role works perfectly fine except it assumes you dedicate a server for this purpose, and this does not scale well for multiple clusters.

What I really need is an on-demand client; only when I need to do manage ceph.

Once ceph.conf is in fetch directory, I can safely push them to git along with keyrings.

**Comment by andymcc** [comment]
Ok I see, so the plan is to push it to git to use at some other point. Thanks for explaining the change to me.
The change seems fine, but I'd like to see a toggle variable that defaults to false.
"ceph_conf_local" or similar, that way it won't generate the conf locally for normal use cases but can be set to configure it when the variable is set to true.

**Comment by guits** [comment]
@andymcc what's the concern by generating the file locally in any case?
we are already bound to the fetch directory, I would say it doesn't matter if you have the ceph.conf inside in addition?

**Comment by guits** [comment]
@andymcc since it doesn't seem that getting rid of that fetch directory is something ongoing I wasn't really sure of the benefits of what you suggested, otherwise I agree with what you said

**Comment by haphan** [comment]
Hmm looks like `{{ fsid }}/etc/ceph` is not created. 

Shall we add a task to ensure it is created?

**Comment by leseb** [comment]
@haphan yes please, along with a variable `ceph_conf_local` just like @andymcc suggested.
Thanks.

**Comment by haphan** [comment]
@leseb @guits Please review

**Comment by haphan** [comment]
@guits What is the expected outcome of running `./generate_group_vars_sample.sh`?

```bash
[root@dev work]# ./generate_group_vars_sample.sh
[root@dev work]# echo $?
0
```

**Comment by guits** [comment]
@haphan no output expected, you should see a diff using `git diff`

**Comment by haphan** [comment]
Well, no new changes found. I was actually expecting new variable in `group_vars/all.yaml.sample` but it is not the case.

**Comment by haphan** [comment]
@guits my bad. Ignore my previous comment. Pushing an update now.

**Comment by haphan** [comment]
@guits Can you take a look and merge this PR?

**Pull request merged**